### PR TITLE
minor: Unify "blank lines before namespace" fixers

### PR DESCRIPTION
**Pull request by Wirone**
- Introduce `blank_lines_before_namespace` that is more flexible than both current fixers combined
- Deprecate existing fixers and proxy them to the newly introduced fixer
- Improve logic that handles line breaks to handle allowed ranges better (use configured min or max depending on current preceding blank lines count)
- Update rule sets to use new fixer

---

I decided to go with plural form in the rule's name (`lines` instead of `line`) because in current implementation it supports multiple lines (0+). Config options are based on line breaks not on blank lines because it would be impossible to distinguish these 2 cases:

```php
<?php namespace Foo;
```

and

```php
<?php
namespace Foo;
```

In both cases there is 0 blank lines before namespace, and both can be achieved by defining proper min/max for line breaks.

It could be achieved with hard-coded strategies that would simplify config into 1 option, but I think it would be better to keep it flexible - there were feature requests already for more sophisticated strategies (in other fixers), so this approach makes it possible to use any 2+ blank lines before namespace.

### CODE REVIEW

**Comment by Wirone** [comment]
@kubawerlos please take a look even if there's an issue with priorities - I need to look at it, but I'll do it later. I would like to know if this approach (config options) is good 🙂.

File: src/Fixer/NamespaceNotation/BlankLinesBeforeNamespaceFixer.php [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+     */
+    public function getPriority(): int
+    {
+        return 0;
+    }
+
+    protected function createConfigurationDefinition(): FixerConfigurationResolverInterface
+    {
+        return new FixerConfigurationResolver([
+            (new FixerOptionBuilder('min_line_breaks', 'Minimum line breaks that should exist before namespace declaration.'))
```
Comment (kubawerlos) [inline src/Fixer/NamespaceNotation/BlankLinesBeforeNamespaceFixer.php:67]: Why `min_line_breaks` and `max_line_breaks`? Wouldn't be more developer/user friendly to have `min_blank_lines` and `max_blank_lines`?

If `min_line_breaks` and `max_line_breaks` are to stay shouldn't the fixer be named `line_breaks_before_namesce`?

Comment (Wirone) [reply src/Fixer/NamespaceNotation/BlankLinesBeforeNamespaceFixer.php:67]: Reasoning is in the description 😝. In terms of fixer's name: I wanted to keep consistency with current fixers and other fixers related to blank lines.

Comment (kubawerlos) [reply src/Fixer/NamespaceNotation/BlankLinesBeforeNamespaceFixer.php:67]: Weird, but fair enough.

I thought that line with `<?php` is reserved for `declare(strict_types=1);`, but we can't expect that strict types will be used.

File: src/AbstractLinesBeforeNamespaceFixer.php
Comment (kubawerlos) [inline src/AbstractLinesBeforeNamespaceFixer.php:1]: ❤️

**Review by kubawerlos — APPROVED** [review]

File: src/Fixer/NamespaceNotation/BlankLinesBeforeNamespaceFixer.php [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+final class BlankLinesBeforeNamespaceFixer extends AbstractFixer implements WhitespacesAwareFixerInterface, ConfigurableFixerInterface
+{
+    public function getDefinition(): FixerDefinitionInterface
+    {
+        return new FixerDefinition(
+            'Controls blank lines before a namespace declaration.',
+            [
+                new CodeSample("<?php  namespace A {}\n"),
+                new CodeSample("<?php  namespace A {}\n", ['min_line_breaks' => 1]),
+                new CodeSample("<?php\n\n\nnamespace A{}\n", ['max_line_breaks' => 2]),
```
[the remark this answers is not in the document]
Comment (Wirone) [reply src/Fixer/NamespaceNotation/BlankLinesBeforeNamespaceFixer.php]: I've added example with 0 line breaks (namespace in the same line as opening tag), but also examples with `declare(strict_types=1);` and comment between opening tag and the namespace declaration.

**Comment by kubawerlos** [comment]
Thank you @Wirone 👍🏼

File: src/Fixer/NamespaceNotation/CleanNamespaceFixer.php [hunk 3]
```diff
@@ -14,14 +14,14 @@
...
-use PhpCsFixer\AbstractLinesBeforeNamespaceFixer;
+use PhpCsFixer\AbstractFixer;
 use PhpCsFixer\FixerDefinition\FixerDefinition;
 use PhpCsFixer\FixerDefinition\FixerDefinitionInterface;
 use PhpCsFixer\FixerDefinition\VersionSpecification;
 use PhpCsFixer\FixerDefinition\VersionSpecificCodeSample;
 use PhpCsFixer\Tokenizer\Tokens;
 
-final class CleanNamespaceFixer extends AbstractLinesBeforeNamespaceFixer
+final class CleanNamespaceFixer extends AbstractFixer
```
Comment (keradus) [inline src/Fixer/NamespaceNotation/CleanNamespaceFixer.php:24]: wow

**Pull request merged**